src/Compiler/Utilities/Caches.fs

```diff
@@ -29,29 +29,27 @@ type CacheOptions =
 [<Sealed; NoComparison; NoEquality>]
 [<DebuggerDisplay("{ToString()}")>]
 type CachedEntity<'Key, 'Value> =
-    val mutable Key: 'Key
-    val mutable Value: 'Value
-    val mutable AccessCount: int64
-    val mutable Node: LinkedListNode<CachedEntity<'Key, 'Value>> voption
+    val mutable private key: 'Key
+    val mutable private value: 'Value
 
-    new(key, value) =
-        {
-            Key = key
-            Value = value
-            AccessCount = 0L
-            Node = ValueNone
-        }
+    [<DefaultValue(false)>]
+    val mutable private node: LinkedListNode<CachedEntity<'Key, 'Value>>
 
-    // This is one time initialization, outside of the constructor because of circular reference.
-    // The contract is that each CachedEntity that the EntityPool produces, has Node assigned.
-    member this.WithNode() =
-        this.Node <- ValueSome(LinkedListNode this)
-        this
+    private new(key, value) = { key = key; value = value }
+
+    member this.Node = this.node
+    member this.Key = this.key
+    member this.Value = this.value
```

Comment:
Make it clear that these things are not intended to be mutable from consumer's pov.